### PR TITLE
Phase2-hgx323 Further correction for ring number assignment of scintillator hits in HGCal geometry versions beyond V13

### PR DESCRIPTION
**Pull request by bsunanda**
#### PR description:

Further correction for ring number assignment of scintillator hits in HGCal geometry versions V14, V15, V16 and V17

#### PR validation:

Use the runTheMatrix test workflows

#### If this PR is a backport please specify the original PR and why you need to backport that PR. If this PR will be backported please specify to which release cycle the backport is meant for:

This is a bug fix for all  versions of HGCal geometry

### CODE REVIEW

**Comment by bsunanda** [comment]
@cmsbuild Please test

**Comment by srimanob** [comment]
type bug-fix

**Comment by srimanob** [comment]
Hi @bsunanda 
Could you please elaborate more detail of the bug and this fix? This will help us to understand the situation more clearly.

- Does the bug come after 12_4_0_pre4 as discussed in validation result, or it was there before?
- If it was there before, how it is found? Do we have any kind of DQM that can spot on this? So that when new relvals come, validator can check on it.

Thanks.

**Comment by bsunanda** [comment]
This bug was there all the time. When the position at the border of 2 rings, it is usually referred to as the lower ring number. This is fine as long as there is no tile in the lower ring. We did not have a good valid test of a detid earlier. We have to introduce one now and that revealed the discrepancy. We do have detailed testing code in the repository and we do conduct them to look into details. The standard validation code can look into broad discrepancies so far and the detailed tests are better done by the experts of the geometry code.

**Comment by civanch** [comment]
+1

**Comment by srimanob** [comment]
+Upgrade

The fix should be followed up in validation.

File: Geometry/HGCalCommonData/src/HGCalDDDConstants.cc [hunk 1]
```diff
@@ -590,13 +590,14 @@ bool HGCalDDDConstants::isValidHex8(int layer, int modU, int modV, int cellU, in
...
+bool HGCalDDDConstants::isValidTrap(int zside, int layer, int irad, int iphi) const {
   // Check validity for a layer|eta|phi of scintillator
   const auto& indx = getIndex(layer, true);
   if (indx.first < 0)
     return false;
-  return ((irad >= hgpar_->iradMinBH_[indx.first]) && (irad <= (hgpar_->iradMaxBH_[indx.first] + 1)) && (iphi > 0) &&
-          (iphi <= hgpar_->scintCells(layer)));
+  bool ok = ((irad >= hgpar_->iradMinBH_[indx.first]) && (irad <= (hgpar_->iradMaxBH_[indx.first] + 1)) && (iphi > 0) &&
+             (iphi <= hgpar_->scintCells(layer)));
+  return ((ok && trapezoidFile()) ? tileExist(zside, layer, irad, iphi) : ok);
```
Comment (perrotta) [inline Geometry/HGCalCommonData/src/HGCalDDDConstants.cc:600]: To be correct, it should be
```suggestion
  return ((ok && trapezoidFile()) ? tileExist(zside, layer, irad, iphi) : true);
```
(even if by  now the net result is the same)

**Comment by perrotta** [comment]
+1
- Let have the bug fix in pre5. The incorrectness in https://github.com/cms-sw/cmssw/pull/39164#pullrequestreview-1085751740 should preferably get fixed in a follow up PR, though
- Quite several changes in the  2026D77 35034 wf: as written in https://github.com/cms-sw/cmssw/pull/39164#issuecomment-1227440841, it must be followed up in rhe validation

**Comment by srimanob** [comment]
By the way, I see failure in Static Check build log,
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9f1384/27040/llvm-analysis/runStaticChecks.log

Sorry that I miss to look on it when approved the PR.

**Comment by srimanob** [comment]
> +1
> 
> * Let have the bug fix in pre5. The incorrectness in [Phase2-hgx323 Further correction for ring number assignment of scintillator hits in HGCal geometry versions V16 and V17 #39164 (review)](https://github.com/cms-sw/cmssw/pull/39164#pullrequestreview-1085751740) should preferably get fixed in a follow up PR, though
> * Quite several changes in the  2026D77 35034 wf: as written in [Phase2-hgx323 Further correction for ring number assignment of scintillator hits in HGCal geometry versions V16 and V17 #39164 (comment)](https://github.com/cms-sw/cmssw/pull/39164#issuecomment-1227440841), it must be followed up in rhe validation

Hmm ... D77 is not expected for the change. There is nothing in PR mentioned about the effect on old geometry. I think this question should be answered first.

**Comment by perrotta** [comment]
Indeed @srimanob , we were too quick

Let revert this, so that it won't pollute pre5

@bsunanda please redo the PR, and fix it by taking https://github.com/cms-sw/cmssw/pull/39164#issuecomment-1227502998 and https://github.com/cms-sw/cmssw/pull/39164#issuecomment-1227506017 into account

File: Geometry/HGCalCommonData/src/HGCalDDDConstants.cc (hunk 1 above)
Comment (bsunanda) [reply Geometry/HGCalCommonData/src/HGCalDDDConstants.cc:600]: I don't think your correction is right - irrespective of trapezoidFile() the earlier test is good for all types of scintillators

**Comment by bsunanda** [comment]
Please do not revert it. The static errors in topology is for debug. I shall take care of that as well as for HCAL stuff separately.

**Comment by srimanob** [comment]
Hi @bsunanda 
Thanks for clarification. Please open the new PR.

**Comment by bsunanda** [comment]
Why a new PR with the same content - cannot be this one restored back?
________________________________

**Comment by srimanob** [comment]
I leave the technical part to @cms-sw/orp-l2 since this PR is merged already. I don't know if it is possible to do.

File: Geometry/HGCalCommonData/src/HGCalDDDConstants.cc (hunk 1 above)
Comment (perrotta) [reply Geometry/HGCalCommonData/src/HGCalDDDConstants.cc:600]: > I don't think your correction is right - irrespective of trapezoidFile() the earlier test is good for all types of scintillators

I understand that in practice `ok && ok` is identical to `ok && true`. But translating in C++ the sentence "irrespective of trapezoidFile() the earlier test is good for all types of scintillators" is exactly `ok && true`

**Pull request merged**